crates/noirc_frontend/src/monomorphisation/mod.rs

```diff
@@ -0,0 +1,576 @@
+            HirPattern::Tuple(patterns, _) => {
+                let fields = unwrap_tuple_type(typ);
+                self.unpack_tuple_pattern(value, patterns.into_iter().zip(fields))
+            }
+            HirPattern::Struct(_, patterns, _) => {
+                let fields = unwrap_struct_type(typ);
+                let patterns = patterns.into_iter().map(|(ident, pattern)| {
+                    let typ = fields[&ident.0.contents].clone();
+                    (pattern, typ)
+                });
+                self.unpack_tuple_pattern(value, patterns)
+            }
+        }
+    }
+
+    fn unpack_tuple_pattern(
+        &mut self,
+        value: ast::Expression,
+        fields: impl Iterator<Item = (HirPattern, HirType)>,
+    ) -> ast::Expression {
+        let fresh_id = self.next_definition_id();
+
+        let mut definitions = vec![ast::Expression::Let(ast::Let {
+            id: fresh_id,
+            name: "_".into(),
+            expression: Box::new(value),
+        })];
+
+        for (i, (field_pattern, field_type)) in fields.into_iter().enumerate() {
+            let typ = self.convert_type(&field_type);
+            let name = i.to_string();
+            let new_rhs =
+                ast::Expression::Ident(ast::Ident { location: None, id: fresh_id, name, typ });
+            let new_rhs = ast::Expression::ExtractTupleField(Box::new(new_rhs), i);
+            let new_expr = self.unpack_pattern(field_pattern, new_rhs, &field_type);
+            definitions.push(new_expr);
+        }
+
+        ast::Expression::Block(definitions)
+    }
+
+    fn ident(&mut self, ident: HirIdent) -> ast::Ident {
+        let id = self.lookup_local(ident.id).unwrap();
+        let name = self.interner.definition_name(ident.id).to_owned();
+        let typ = self.convert_type(&self.interner.id_type(ident.id));
+        ast::Ident { location: Some(ident.location), id, name, typ }
+    }
+
+    /// Convert a non-tuple/struct type to a monomorphised type
+    fn convert_type(&self, typ: &HirType) -> ast::Type {
+        match typ {
+            HirType::FieldElement(_) => ast::Type::Field,
+            HirType::Integer(_, sign, bits) => ast::Type::Integer(*sign, *bits),
+            HirType::Bool(_) => ast::Type::Bool,
+            HirType::Unit => ast::Type::Unit,
+
+            HirType::Array(size, element) => {
+                let size = size.array_length().unwrap_or(0);
+                let element = self.convert_type(element.as_ref());
+                ast::Type::Array(size, Box::new(element))
+            }
+
+            HirType::PolymorphicInteger(_, binding)
+            | HirType::TypeVariable(binding)
+            | HirType::NamedGeneric(binding, _) => {
+                if let TypeBinding::Bound(binding) = &*binding.borrow() {
+                    return self.convert_type(binding);
+                }
+                // Default any remaining unbound type variables to Field
```

Comment:
This is necessary for defaulting PolymorphicInteger to Field, though I can separate out the other two cases of NamedGeneric and TypeVariable. I see no reason to issue an error to the user though since the only way defaulting unbound type variables can change behavior is if the compiler does an automatic impl search for traits - which we don't.

Reply:
Are you saying unbound variable are not getting there? In that case we should have `unreachable!()`

Reply:
Thank you for your explanation, I agree now it is ok not to generate an error, however I think we should have an Error type and use this one instead of Field here. I understand the variable won't be used so it will not make a difference but I think it is a safeguard in case of future additions/features.

Reply:
I'll add a comment on it - I wan to avoid an error type for now since it could show up in situations like arrays of size 0 (if we ever allow them) which would be of type `[T]`. Since the outer array type and handling the array itself would be valid (only valid operation would be taking its length) panicking when the ssa pass sees a `[Error]` would also be too restrictive.